src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs

```diff
@@ -520,6 +521,17 @@ private SignatureProvider CreateSignatureProvider(SecurityKey key, string algori
             return signatureProvider;
         }
 
+        /// <summary>
+        /// For some security key types, in some runtimes, it's not possible to extract public key material and create an <see cref="SecurityKey.InternalId"/>.
+        /// In these cases, <see cref="SecurityKey.InternalId"/> will be an empty string, and these keys should not be cached.
+        /// </summary>
+        /// <param name="signatureProvider"><see cref="SignatureProvider"/> to be examined.</param>
+        /// <returns><c>True</c> if <paramref name="signatureProvider"/> should be cached, <c>false</c> otherwise.</returns>
+        internal static bool ShouldCacheSignatureProvider(SignatureProvider signatureProvider)
```

Comment:
I've moved this method out of the cache implementation after discussing an associated issue with George.

In the case where a SignatureProvider has an empty string as an internal ID, it should **never** be cached, regardless of the caching implementation.